### PR TITLE
Production ready docker deployment + OAuth authority

### LINKED ISSUE #870: Failed to build Docker Image for athenz-cli-util.
Hi, I tried the latest master branch to build Athenz Docker Images.
But it didn't work.

```
$ git rev-parse --short HEAD
4e7863d8

$ cd `git rev-parse --show-toplevel`/docker

$ make build | tee ./athenz-docker-build.log
```

Since `athenz-cert-refresher` not found, mvn install failed for `athenz-zts-java-client-core`.

```
[INFO] ------------< com.yahoo.athenz:athenz-zts-java-client-core >------------
[INFO] Building athenz-zts-java-client-core 1.8.45-SNAPSHOT              [6/23]
[INFO] --------------------------------[ jar ]---------------------------------
[WARNING] The POM for com.yahoo.athenz:athenz-cert-refresher:jar:1.8.45-SNAPSHOT is missing, no dependency information available
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for rdl-gen-athenz-go-client 1.8.45-SNAPSHOT:
[INFO]
[INFO] rdl-gen-athenz-go-client ........................... SUCCESS [ 31.582 s]
[INFO] athenz-zms-core .................................... SUCCESS [ 16.657 s]
[INFO] athenz-zts-core .................................... SUCCESS [  6.829 s]
[INFO] athenz-auth-core ................................... SUCCESS [  8.331 s]
[INFO] athenz-client-common ............................... SUCCESS [  0.518 s]
[INFO] athenz-zts-java-client-core ........................ FAILURE [  0.114 s]
[INFO] athenz-zts-java-client ............................. SKIPPED
[INFO] zms-go-client ...................................... SKIPPED
[INFO] zts-go-client ...................................... SKIPPED
[INFO] zmscli ............................................. SKIPPED
[INFO] zmssvctoken ........................................ SKIPPED
[INFO] ztsclientutil ...................................... SKIPPED
[INFO] athenzconf ......................................... SKIPPED
[INFO] zms-cli ............................................ SKIPPED
[INFO] athenz-conf ........................................ SKIPPED
[INFO] zms-svctoken ....................................... SKIPPED
[INFO] athenz-zpe-policy-updater .......................... SKIPPED
[INFO] zpe-updater ........................................ SKIPPED
[INFO] zts-roletoken ...................................... SKIPPED
[INFO] zts-accesstoken .................................... SKIPPED
[INFO] zts-rolecert ....................................... SKIPPED
[INFO] zts-svccert ........................................ SKIPPED
[INFO] athenz-utils ....................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:04 min
[INFO] Finished at: 2020-01-29T06:22:44Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project athenz-zts-java-client-core: Could not resolve dependencies for project com.yahoo.athenz:athenz-zts-java-client-core:jar:1.8.45-SNAPSHOT: Could not find artifact com.yahoo.athenz:athenz-cert-refresher:jar:1.8.45-SNAPSHOT -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :athenz-zts-java-client-core
The command '/bin/sh -c GO111MODULE=on   CGO_ENABLED=0   CGO_CXXFLAGS="-g -Ofast -march=native"   CGO_FFLAGS="-g -Ofast -march=native"   CGO_LDFLAGS="-g -Ofast -march=native"   GOOS=$(go env GOOS)   GOARCH=$(go env GOARCH)   mvn install -Dmaven.test.skip=true   -pl core/zms   -pl core/zts   -pl libs/java/auth_core   -pl libs/java/client_common   -pl clients/java/zts/core   -pl clients/java/zts/shade   -pl rdl/rdl-gen-athenz-go-client   -pl clients/go/zms   -pl clients/go/zts   -pl libs/go/zmscli   -pl libs/go/zmssvctoken   -pl libs/go/athenzutils   -pl libs/go/athenzconf   -pl utils/zms-cli   -pl utils/athenz-conf   -pl utils/zms-svctoken   -pl utils/zpe_policy_updater   -pl utils/zpe-updater   -pl utils/zts-roletoken   -pl utils/zts-accesstoken   -pl utils/zts-rolecert   -pl utils/zts-svccert   -pl assembly/utils' returned a non-zero code: 1
make: *** [build] Error 1
```

In my environment, I could resolve the above error with below fixes. ( I can send a PR)

```diff
diff --git a/docker/util/Dockerfile b/docker/util/Dockerfile
index 3ef90ef6..031f86bd 100644
--- a/docker/util/Dockerfile
+++ b/docker/util/Dockerfile
@@ -23,6 +23,7 @@ RUN GO111MODULE=on \
   -pl core/zts \
   -pl libs/java/auth_core \
   -pl libs/java/client_common \
+  -pl libs/java/cert_refresher \
   -pl clients/java/zts/core \
   -pl clients/java/zts/shade \
   # go projects
```

### PR DESCRIPTION
**Pull request by WindzCUHK**
## concerns

1. Is there official athenz logo (zms, zts) to be used inside diagrams for documentation?
2. Plx help to update the README.md, when the tag version of merging this PR is known. https://github.com/yahoo/athenz/pull/940#discussion_r411223839

## changes

1. rework docker deployment, make it production ready
	- related files:
		- `docker/*`
		- `servers/zts/src/main/java/com/yahoo/athenz/zts/cert/impl/KeyStoreCertSigner.java`
1. add OAuth authority
	- refactor:
		- `libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/impl/CertificateAuthority.java`
		- `libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/AuthorityConsts.java`
		- `libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/util/AthenzUtils.java`
	- related files:
		- `libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/oauth/OAuthCertBoundJwtAccessTokenAuthority.java`
		- `libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/oauth/*`
1. update jjwt.version
	- related files:
		- `libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/token/AccessToken.java`
		- `libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/token/IdToken.java`
		- `libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/token/OAuth2Token.java`
		- `libs/java/auth_core/pom.xml`
		- `servers/zts/pom.xml`
		- `pom.xml`
1. support Auth0 as Athenz IdP
	- related files:
		- `docker/docs/IdP/Auth0.md`
		- `contributions/authority/auth0/*`
1. fix dead link
	- related files:
		- `servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSConsts.java`
		- `servers/zts/src/main/java/com/yahoo/athenz/zts/ZTSConsts.java`
1. remove bouncer
	- related files:
		- `aws-setup/zms-setup/build/conf/zms.properties`
1. fix space
	- related files:
		- `libs/java/server_common/src/test/java/com/yahoo/athenz/common/server/log/impl/AuditLogMsgBuilderTest.java`

## details for docker rework

1. use `CertificateAuthority` (user certificate) instead of `UserAuthority` during bootstrap
1. re-organize scripts
1. add docker image for setup and testing
1. minimize cli dependency
1. add admin user in DB instead of using root
1. add linux user inside docker images
1. remove docker stack
1. add healthcheck and labels in docker images
1. change JAVA base image to [adoptopenjdk/openjdk8](https://hub.docker.com/r/adoptopenjdk/openjdk8)
1. add OAuth authority and Auth0 as example
1. [skip] provide CSR template for requesting certificate from public CA
	- tls config varies vigorously between companies, it is out of Athenz's scope

### close
- close #870 

> I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by WindzCUHK** [comment]
Just in case you also want to verify the Auth0 package, plx follow the below steps.
https://github.com/WindzCUHK/athenz/pull/2#issuecomment-595079197

File: docker/README.md [hunk 1]
```diff
@@ -6,117 +6,157 @@
...
 cd `git rev-parse --show-toplevel`/docker
 
-# there are a lot of build logs. You may want to check it inside a log file later on.
-make build | tee ./athenz-docker-build.log
+# it takes about 15-30 mins
+make build
+
+# TODO: update the min. version before merge
+# P.S. the latest code may cause docker build to fail, please use older version by specifying the tag version (> v1.8.60) or post an issue
+# make build TAG=v1.8.60
```
Comment (WindzCUHK) [inline docker/README.md:41]: Plx help to update the README.md, when the tag version of merging this PR is known.

**Comment by havetisyan** [comment]
@WindzCUHK there are way too many changes here I would suggest to break them up into separate PRs. It would be easier to review and merge.

**Comment by WindzCUHK** [comment]
> @WindzCUHK there are way too many changes here I would suggest to break them up into separate PRs. It would be easier to review and merge.

Noted. We break to separate PRs.

**Pull request closed without merging**